// stryker/src/isolated-runner/TestRunnerDecorator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var TestRunnerDecorator = /** @class */ (function () {
    function TestRunnerDecorator(testRunnerProducer) {
        this.testRunnerProducer = testRunnerProducer;
        this.createInnerRunner();
    }
    TestRunnerDecorator.prototype.init = function () {
        if (this.innerRunner.init) {
            return this.innerRunner.init() || Promise.resolve();
        }
        else {
            return Promise.resolve();
        }
    };
    TestRunnerDecorator.prototype.createInnerRunner = function () {
        this.innerRunner = this.testRunnerProducer();
    };
    TestRunnerDecorator.prototype.run = function (options) {
        return this.innerRunner.run(options);
    };
    TestRunnerDecorator.prototype.dispose = function () {
        if (this.innerRunner.dispose) {
            return this.innerRunner.dispose() || Promise.resolve();
        }
        else {
            return Promise.resolve();
        }
    };
    return TestRunnerDecorator;
}());
exports.default = TestRunnerDecorator;
