// stryker-javascript-mutator/src/mutators/StringLiteralMutator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var babel_core_1 = require("babel-core");
var StringLiteralMutator = /** @class */ (function () {
    function StringLiteralMutator() {
        this.name = 'StringLiteral';
    }
    StringLiteralMutator.prototype.mutate = function (node, copy) {
        var nodes = [];
        if (babel_core_1.types.isTemplateLiteral(node)) {
            var mutatedNode = {
                type: 'StringLiteral',
                start: node.start,
                end: node.end,
                loc: node.loc,
                value: ''
            };
            if (node.quasis.length === 1 && node.quasis[0].value.raw.length === 0) {
                mutatedNode.value = 'Stryker was here!';
            }
            nodes.push(mutatedNode);
        }
        else if ((!node.parent || (!babel_core_1.types.isImportDeclaration(node.parent) && !babel_core_1.types.isJSXAttribute(node.parent)))
            && babel_core_1.types.isStringLiteral(node)) {
            var mutatedNode = copy(node);
            mutatedNode.value = mutatedNode.value.length === 0 ? 'Stryker was here!' : '';
            nodes.push(mutatedNode);
        }
        return nodes;
    };
    return StringLiteralMutator;
}());
exports.default = StringLiteralMutator;
