// stryker-api/src/report/MutantStatus.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var MutantStatus;
(function (MutantStatus) {
    /**
     * The status of a survived mutant, because it was not covered by any test.
     */
    MutantStatus[MutantStatus["NoCoverage"] = 0] = "NoCoverage";
    /**
     * The status of a killed mutant.
     */
    MutantStatus[MutantStatus["Killed"] = 1] = "Killed";
    /**
     * The status of a survived mutant.
     */
    MutantStatus[MutantStatus["Survived"] = 2] = "Survived";
    /**
     * The status of a timed out mutant.
     */
    MutantStatus[MutantStatus["TimedOut"] = 3] = "TimedOut";
    /**
     * The status of a mutant of which the tests resulted in a runtime error.
     *
     * For example: the following piece of javascript code will result in a runtime error:
     *
     * ```javascript
     * const fs = require('f' - 's'); // mutated code
     * ```
     *
     * Mutants that result in a runtime error are not taken into account during score calculation.
     */
    MutantStatus[MutantStatus["RuntimeError"] = 4] = "RuntimeError";
    /**
     * The status of a mutant which could not be transpiled.
     * For example: the following piece of typescript code will give a TranspileError:
     *
     * ```typescript
     * const a: 5 = 0; // mutated code
     * ```
     *
     * Mutants that result in a TranspileError are not taken into account during score calculation.
     */
    MutantStatus[MutantStatus["TranspileError"] = 5] = "TranspileError";
})(MutantStatus || (MutantStatus = {}));
exports.default = MutantStatus;
