// stryker/src/ConfigValidator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var log4js_1 = require("log4js");
var ConfigValidator = /** @class */ (function () {
    function ConfigValidator(strykerConfig, testFramework) {
        this.strykerConfig = strykerConfig;
        this.testFramework = testFramework;
        this.isValid = true;
        this.log = log4js_1.getLogger(ConfigValidator.name);
    }
    ConfigValidator.prototype.validate = function () {
        this.validateTestFramework();
        this.validateThresholds();
        this.validateMutator();
        this.validateLogLevel();
        this.validateTimeout();
        this.validateIsNumber('port', this.strykerConfig.port);
        this.validateIsNumber('maxConcurrentTestRunners', this.strykerConfig.maxConcurrentTestRunners);
        this.validateIsStringArray('plugins', this.strykerConfig.plugins);
        this.validateIsStringArray('reporter', this.strykerConfig.reporter);
        this.validateIsStringArray('transpilers', this.strykerConfig.transpilers);
        this.validateCoverageAnalysis();
        this.validateCoverageAnalysisWithRespectToTranspilers();
        this.crashIfNeeded();
    };
    ConfigValidator.prototype.validateTestFramework = function () {
        if (this.strykerConfig.coverageAnalysis === 'perTest' && !this.testFramework) {
            this.invalidate('Configured coverage analysis "perTest" requires there to be a testFramework configured. Either configure a testFramework or set coverageAnalysis to "all" or "off".');
        }
    };
    ConfigValidator.prototype.validateMutator = function () {
        var mutator = this.strykerConfig.mutator;
        if (typeof mutator === 'object') {
            var mutatorDescriptor = mutator;
            this.validateIsString('mutator.name', mutatorDescriptor.name);
            this.validateIsStringArray('mutator.excludedMutations', mutatorDescriptor.excludedMutations);
        }
        else if (typeof mutator !== 'string') {
            this.invalidate("Value \"" + mutator + "\" is invalid for `mutator`. Expected either a string or an object");
        }
    };
    ConfigValidator.prototype.validateThresholds = function () {
        var thresholds = this.strykerConfig.thresholds;
        this.validateThresholdsValueExists('high', thresholds.high);
        this.validateThresholdsValueExists('low', thresholds.low);
        this.validateThresholdValue('high', thresholds.high);
        this.validateThresholdValue('low', thresholds.low);
        this.validateThresholdValue('break', thresholds.break);
        if (thresholds.high < thresholds.low) {
            this.invalidate("`thresholds.high` is lower than `thresholds.low` (" + thresholds.high + " < " + thresholds.low + ")");
        }
    };
    ConfigValidator.prototype.validateThresholdValue = function (name, value) {
        if (typeof value === 'number' && (value < 0 || value > 100)) {
            this.invalidate("Value \"" + value + "\" is invalid for `thresholds." + name + "`. Expected a number between 0 and 100");
        }
    };
    ConfigValidator.prototype.validateThresholdsValueExists = function (name, value) {
        if (typeof value !== 'number') {
            this.invalidate("Value \"" + value + "\" is invalid for `thresholds." + name + "`. Expected a number between 0 and 100");
        }
    };
    ConfigValidator.prototype.validateLogLevel = function () {
        var logLevel = this.strykerConfig.logLevel;
        var VALID_LOG_LEVEL_VALUES = ['fatal', 'error', 'warn', 'info', 'debug', 'trace', 'all', 'off'];
        if (VALID_LOG_LEVEL_VALUES.indexOf(logLevel) < 0) {
            this.invalidate("Value \"" + logLevel + "\" is invalid for `logLevel`. Expected one of the following: " + this.joinQuotedList(VALID_LOG_LEVEL_VALUES));
        }
    };
    ConfigValidator.prototype.validateTimeout = function () {
        this.validateIsNumber('timeoutMs', this.strykerConfig.timeoutMs);
        this.validateIsNumber('timeoutFactor', this.strykerConfig.timeoutFactor);
    };
    ConfigValidator.prototype.validateCoverageAnalysis = function () {
        var VALID_COVERAGE_ANALYSIS_VALUES = ['perTest', 'all', 'off'];
        var coverageAnalysis = this.strykerConfig.coverageAnalysis;
        if (VALID_COVERAGE_ANALYSIS_VALUES.indexOf(coverageAnalysis) < 0) {
            this.invalidate("Value \"" + coverageAnalysis + "\" is invalid for `coverageAnalysis`. Expected one of the following: " + this.joinQuotedList(VALID_COVERAGE_ANALYSIS_VALUES));
        }
    };
    ConfigValidator.prototype.validateCoverageAnalysisWithRespectToTranspilers = function () {
        if (Array.isArray(this.strykerConfig.transpilers) &&
            this.strykerConfig.transpilers.length > 1 &&
            this.strykerConfig.coverageAnalysis !== 'off') {
            this.invalidate("Value \"" + this.strykerConfig.coverageAnalysis + "\" for `coverageAnalysis` is invalid with multiple transpilers (configured transpilers: " + this.strykerConfig.transpilers.join(', ') + "). Please report this to the Stryker team if you whish this feature to be implemented");
        }
    };
    ConfigValidator.prototype.crashIfNeeded = function () {
        if (!this.isValid) {
            process.exit(1);
        }
    };
    ConfigValidator.prototype.validateIsNumber = function (fieldName, value) {
        if (typeof value !== 'number') {
            this.invalidate("Value \"" + value + "\" is invalid for `" + fieldName + "`. Expected a number");
        }
    };
    ConfigValidator.prototype.validateIsString = function (fieldName, value) {
        if (typeof value !== 'string') {
            this.invalidate("Value \"" + value + "\" is invalid for `" + fieldName + "`. Expected a string");
        }
    };
    ConfigValidator.prototype.validateIsStringArray = function (fieldName, value) {
        var _this = this;
        if (!Array.isArray(value)) {
            this.invalidate("Value \"" + value + "\" is invalid for `" + fieldName + "`. Expected an array");
        }
        else {
            value.forEach(function (v) {
                if (typeof v !== 'string') {
                    _this.invalidate("Value \"" + v + "\" is an invalid element of `" + fieldName + "`. Expected a string");
                }
            });
        }
    };
    ConfigValidator.prototype.invalidate = function (message) {
        this.log.fatal(message);
        this.isValid = false;
    };
    ConfigValidator.prototype.joinQuotedList = function (arr) {
        return arr.map(function (v) { return "\"" + v + "\""; }).join(', ');
    };
    return ConfigValidator;
}());
exports.default = ConfigValidator;
