// stryker/src/PluginLoader.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var fs = require("mz/fs");
var path = require("path");
var log4js_1 = require("log4js");
var _ = require("lodash");
var fileUtils_1 = require("./utils/fileUtils");
var IGNORED_PACKAGES = ['stryker-cli', 'stryker-api'];
var PluginLoader = /** @class */ (function () {
    function PluginLoader(plugins) {
        this.plugins = plugins;
        this.log = log4js_1.getLogger(PluginLoader.name);
    }
    PluginLoader.prototype.load = function () {
        var _this = this;
        this.getModules().forEach(function (moduleName) { return _this.requirePlugin(moduleName); });
    };
    PluginLoader.prototype.getModules = function () {
        var _this = this;
        var modules = [];
        this.plugins.forEach(function (pluginExpression) {
            if (_.isString(pluginExpression)) {
                if (pluginExpression.indexOf('*') !== -1) {
                    // Plugin directory is the node_modules folder of the module that installed stryker
                    // So if current __dirname is './stryker/src' than the plugin directory should be 2 directories above
                    var pluginDirectory_1 = path.normalize(__dirname + '/../..');
                    var regexp_1 = new RegExp('^' + pluginExpression.replace('*', '.*'));
                    _this.log.debug('Loading %s from %s', pluginExpression, pluginDirectory_1);
                    var plugins = fs.readdirSync(pluginDirectory_1)
                        .filter(function (pluginName) { return IGNORED_PACKAGES.indexOf(pluginName) === -1 && regexp_1.test(pluginName); })
                        .map(function (pluginName) { return pluginDirectory_1 + '/' + pluginName; });
                    if (plugins.length === 0) {
                        _this.log.debug('Expression %s not resulted in plugins to load', pluginExpression);
                    }
                    plugins
                        .map(function (plugin) { return path.basename(plugin); })
                        .map(function (plugin) {
                        _this.log.debug('Loading plugins %s (matched with expression %s)', plugin, pluginExpression);
                        return plugin;
                    })
                        .forEach(function (p) { return modules.push(p); });
                }
                else {
                    modules.push(pluginExpression);
                }
            }
            else {
                _this.log.warn('Ignoring plugin %s, as its not a string type', pluginExpression);
            }
        });
        return modules;
    };
    PluginLoader.prototype.requirePlugin = function (name) {
        this.log.debug("Loading plugins " + name);
        try {
            fileUtils_1.importModule(name);
        }
        catch (e) {
            if (e.code === 'MODULE_NOT_FOUND' && e.message.indexOf(name) !== -1) {
                this.log.warn('Cannot find plugin "%s".\n  Did you forget to install it ?\n' +
                    '  npm install %s --save-dev', name, name);
            }
            else {
                this.log.warn('Error during loading "%s" plugin:\n  %s', name, e.message);
            }
        }
    };
    return PluginLoader;
}());
exports.default = PluginLoader;
