// stryker-api/src/config/Config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Config = /** @class */ (function () {
    function Config() {
        this.logLevel = 'info';
        this.timeoutMs = 5000;
        this.timeoutFactor = 1.5;
        this.plugins = ['stryker-*'];
        this.port = 9234;
        this.reporter = ['progress', 'clear-text'];
        this.coverageAnalysis = 'perTest';
        this.mutator = 'es5';
        this.transpilers = [];
        this.maxConcurrentTestRunners = Infinity;
        this.symlinkNodeModules = true;
        this.thresholds = {
            high: 80,
            low: 60,
            break: null
        };
    }
    Config.prototype.set = function (newConfig) {
        var _this = this;
        if (newConfig) {
            Object.keys(newConfig).forEach(function (key) {
                if (typeof newConfig[key] !== 'undefined') {
                    _this[key] = newConfig[key];
                }
            });
        }
    };
    return Config;
}());
exports.default = Config;
