// stryker/src/initializer/NpmClient.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var RestClient_1 = require("typed-rest-client/RestClient");
var log4js_1 = require("log4js");
var objectUtils_1 = require("../utils/objectUtils");
var BASE_NPM_SEARCH = 'https://api.npms.io';
var BASE_NPM_PACKAGE = 'https://registry.npmjs.org';
var getName = function (packageName) {
    return packageName.split('-')[1];
};
var mapSearchResultToPromptOption = function (searchResults) { return searchResults.results.map(function (result) { return ({
    name: getName(result.package.name),
    npm: result.package.name
}); }); };
var handleResult = function (from) { return function (response) {
    if (response.statusCode === 200 && response.result) {
        return response.result;
    }
    else {
        throw new Error("Query " + from + " resulted in http status code: " + response.statusCode + ".");
    }
}; };
var NpmClient = /** @class */ (function () {
    function NpmClient(searchClient, packageClient) {
        if (searchClient === void 0) { searchClient = new RestClient_1.RestClient('npmSearch', BASE_NPM_SEARCH); }
        if (packageClient === void 0) { packageClient = new RestClient_1.RestClient('npm', BASE_NPM_PACKAGE); }
        this.searchClient = searchClient;
        this.packageClient = packageClient;
        this.log = log4js_1.getLogger(NpmClient.name);
    }
    NpmClient.prototype.getTestRunnerOptions = function () {
        return this.search('/v2/search?q=keywords:stryker-test-runner')
            .then(mapSearchResultToPromptOption);
    };
    NpmClient.prototype.getTestFrameworkOptions = function (testRunnerFilter) {
        return this.search('/v2/search?q=keywords:stryker-test-framework')
            .then(function (searchResult) {
            if (testRunnerFilter) {
                searchResult.results = searchResult.results.filter(function (framework) { return framework.package.keywords.indexOf(testRunnerFilter) >= 0; });
            }
            return searchResult;
        })
            .then(mapSearchResultToPromptOption);
    };
    NpmClient.prototype.getMutatorOptions = function () {
        return this.search('/v2/search?q=keywords:stryker-mutator')
            .then(mapSearchResultToPromptOption);
    };
    NpmClient.prototype.getTranspilerOptions = function () {
        return this.search('/v2/search?q=keywords:stryker-transpiler')
            .then(mapSearchResultToPromptOption);
    };
    NpmClient.prototype.getTestReporterOptions = function () {
        return this.search("/v2/search?q=keywords:stryker-reporter")
            .then(mapSearchResultToPromptOption);
    };
    NpmClient.prototype.getAdditionalConfig = function (packageName) {
        var _this = this;
        return this.packageClient.get("/" + packageName + "/latest")
            .then(handleResult(BASE_NPM_PACKAGE + "/" + packageName))
            .then(function (pkg) { return pkg.initStrykerConfig || {}; })
            .catch(function (err) {
            _this.log.warn("Could not fetch additional initialization config for dependency " + packageName + ". You might need to configure it manually", err);
            return {};
        });
    };
    NpmClient.prototype.search = function (query) {
        var _this = this;
        var call = BASE_NPM_SEARCH + query;
        this.log.debug("Searching: " + call);
        return this.searchClient.get(query)
            .then(handleResult(call))
            .catch(function (err) {
            _this.log.error("Unable to reach " + BASE_NPM_SEARCH + " (for query " + query + "). Please check your internet connection.", objectUtils_1.errorToString(err));
            var result = {
                total: 0,
                results: []
            };
            return result;
        });
    };
    return NpmClient;
}());
exports.default = NpmClient;
