// stryker/src/transpiler/TranspilerFacade.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var transpile_1 = require("stryker-api/transpile");
var StrykerError_1 = require("../utils/StrykerError");
var NamedTranspiler = /** @class */ (function () {
    function NamedTranspiler(name, transpiler) {
        this.name = name;
        this.transpiler = transpiler;
    }
    return NamedTranspiler;
}());
var TranspilerFacade = /** @class */ (function () {
    function TranspilerFacade(options) {
        this.innerTranspilers = options.config.transpilers
            .map(function (transpilerName) { return new NamedTranspiler(transpilerName, transpile_1.TranspilerFactory.instance().create(transpilerName, options)); });
    }
    TranspilerFacade.prototype.transpile = function (files) {
        return this.performTranspileChain(files);
    };
    TranspilerFacade.prototype.performTranspileChain = function (input, remainingChain) {
        if (remainingChain === void 0) { remainingChain = this.innerTranspilers.slice(); }
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var current, output;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        current = remainingChain.shift();
                        if (!current) return [3 /*break*/, 2];
                        return [4 /*yield*/, current.transpiler.transpile(input)
                                .catch(function (error) {
                                throw new StrykerError_1.default("An error occurred in transpiler \"" + current.name + "\"", error);
                            })];
                    case 1:
                        output = _a.sent();
                        return [2 /*return*/, this.performTranspileChain(output, remainingChain)];
                    case 2: return [2 /*return*/, input];
                }
            });
        });
    };
    return TranspilerFacade;
}());
exports.default = TranspilerFacade;
